src/main/java/frc/robot/Manipulator.java

```diff
@@ -266,38 +268,117 @@ public void startLogging(String basekey) {
 	}
 
 
+	public ManipulatorControl controlManipulator(
+		DoubleSupplier a, DoubleSupplier g, DoubleSupplier w
+	) {
+		return new ManipulatorControl(this, a, g, w);
+	}
+	public ManipulatorControl controlManipulator(
+		DoubleSupplier a, DoubleSupplier g, DoubleSupplier w,
+		BooleanSupplier wr, BooleanSupplier al, BooleanSupplier gl
+	) {
+		return new ManipulatorControl(this, a, g, w, wr, al, gl);
+	}
+
+
 
 
 
-	public static class TestManipulator extends CommandBase {
+	public static class ManipulatorControl extends CommandBase {
 
 		public static final double
 			ARM_WINCH_VOLTAGE_SCALE = 5.0,
-			GRAB_CLAW_VOLTAGE_SCALE = 7.0;
+			ARM_WINCH_LOCK_VOLTAGE = 2.5,
+			GRAB_CLAW_VOLTAGE_SCALE = 7.0,
+			GRAB_CLAW_LOCK_VOLTAGE = 2.5,
+			WRIST_ACCUMULATION_RATE_SCALE = 0.01;	// at full throttle, add 0.01 x 50 loops per second = 0.5 per second change [maximum]
 
 		private final Manipulator
 			manipulator;
 		private final DoubleSupplier
-			arm_percent,
-			grab_percent,
-			wrist_set;
-
-		public TestManipulator(Manipulator m, DoubleSupplier a, DoubleSupplier g, DoubleSupplier w) {
+			arm_rate,
+			grab_rate,
+			wrist_rate;
+		private final BooleanSupplier
+			wrist_reset,
+			arm_lock,
+			grab_lock;
+
+		private double
+			wrist_position = Grabber.WRIST_NEUTRAL_PERCENT;
+		private boolean
+			is_arm_locked = false,
+			is_grab_locked = false;
+
+
+		public ManipulatorControl(
+			Manipulator m,
+			DoubleSupplier a, DoubleSupplier g, DoubleSupplier w
+		) {
 			this.manipulator = m;
-			this.arm_percent = a;
-			this.grab_percent = g;
-			this.wrist_set = w;
-		}
+			this.arm_rate = a;
+			this.grab_rate = g;
+			this.wrist_rate = w;
+			this.wrist_reset = this.arm_lock = this.grab_lock = ()->false;
+			super.addRequirements(m.arm, m.grabber);
+		}
+		public ManipulatorControl(
+			Manipulator m,
+			DoubleSupplier a, DoubleSupplier g, DoubleSupplier w,
+			BooleanSupplier wr, BooleanSupplier al, BooleanSupplier gl
+		) {
+			this.manipulator = m;
+			this.arm_rate = a;
+			this.grab_rate = g;
+			this.wrist_rate = w;
+			this.wrist_reset = wr;
+			this.arm_lock = al;
+			this.grab_lock = gl;
+			super.addRequirements(m.arm, m.grabber);
+		}
+
+
```

Comment:
It looks to me like this commented out function is replaced by the code below.  Can the commented code be deleted?

Reply:
This was Zoke's nonfunctional code that you mention in your next comment. I kept it in for reference in case she wants to fix it.